articles/from-theory-to-practice-implementing-totp-2fa-in-python-and-flask/index.md

```diff
@@ -0,0 +1,361 @@
+            <li>Provide the required details (name, secret key).</li>
+            <li>Select time-based authentication.</li>
+            <li>Submit the generated key in the form.</li>
+          </ul>
+        </div>
+        <div class="form-group">
+          <label for="secret">Secret Token</label>
+          <input type="text" class="form-control" id="secret" value="{{ secret }}" readonly>
+        </div>
+        <div class="text-center">
+          <button type="button" class="btn btn-primary" onclick="copySecret()">Copy Secret</button>
+        </div>
+      </form>
+    </div>
+    <div class="col-lg-7">
+      {% with messages = get_flashed_messages(with_categories=true) %}
+      {% if messages %}
+      {% for category, message in messages %}
+      <div class="alert alert-{{ category }}" role="alert">
+        {{ message }}
+      </div>
+      {% endfor %}
+      {% endif %}
+      {% endwith %}
+      <form method="POST">
+        <div class="form-group">
+          <label for="otp">Generated OTP</label>
+          <input type="hidden" name="secret" value="{{ secret }}" required>
+          <input type="number" class="form-control" id="otp" name="otp" required>
+        </div>
+        <div class="text-center">
+          <button type="submit" class="btn btn-primary">Authenticate User</button>
+        </div>
+      </form>
+    </div>
+  </div>
+</div>
+
+<script>
+  function copySecret() {
+    /* Get the text field */
+    var copyText = document.getElementById("secret");
+
+    /* Select the text field */
+    copyText.select();
+    copyText.setSelectionRange(0, 99999); /*For mobile devices*/
+
+    /* Copy the text inside the text field */
+    document.execCommand("copy");
+
+    alert("Successfully copied TOTP secret token!");
+  }
+</script>
+{% endblock %}
+```
+
+![Generating OTP](\engineering-education\from-theory-to-practice-implementing-totp-2fa-in-python-and-flask\Generating OTP.png)
+
+You will also write a route to handle `POST` requests made to the 2FA page and authenticate them. Update the `app.py` file by adding the code below:
+
+```python
+@app.route("/login/2fa/", methods=["POST"])
+def login_2fa_form():
+    secret = request.form.get("secret")
+    otp = int(request.form.get("otp"))
+
+    if pyotp.TOTP(secret).verify(otp):
+        flash("The TOTP 2FA token is valid", "success")
+        return redirect(url_for("login_2fa"))
+    else:
+        flash("You have supplied an invalid 2FA token!", "danger")
+        return redirect(url_for("login_2fa"))
+```
+
+You should get an image similar to the one below when an invalid token is provided:
+
+![invalid OTP](\engineering-education\from-theory-to-practice-implementing-totp-2fa-in-python-and-flask\invalid OTP.png)
+
+You should get an image similar to the one below when a valid token is provided:
+
+![valid OTP](\engineering-education\from-theory-to-practice-implementing-totp-2fa-in-python-and-flask\valid OTP.png)
+
+### Conclusion
+
```

Comment:
Please include a GitHub link of the final code of the application that you built.